File: R/ZeroLines.R
ZeroLines = function(M, V, n_sp, antprob){
  # Check if there are any lines full of zero's in M adjacency matrix
  # Args:
  #  M = Mutualistic adjacency matrix
  #  V = Antagonistic adjacency matrix
  #  n_sp = number os species
  #  antprob = probability of a link become antagonist
  #
  #  Return:
  #    In case of lines full of zeros, make new adjacency matrices
  #
  # Obs: this function works with the Antagonize.R and EndInteraction.R functions.
  #      Make sure to have then in your work directory

  # load packages for new matrices if necessary
  source("Antagonize.R")
  source("EndInteraction.R")
  
  # try several types get a M matrix without lines of zero's
  for(z in 1:1000000){
    m = rowSums(M) # check for zero lines
    
    # make new matrices if there's a line full of zero's in M
    if(any(m == 0)){ 
      M = matrix(1, ncol = n_sp, nrow = n_sp)   # building matrix M (mutualisms)
      diag(M) = 0 # no intraespecific interactions
      
      # Antagonize M
      antagonize = Antagonize(M, antprob)
      M = antagonize[[1]]
      V = antagonize[[2]]
      
      # End antagonism AA
      end = EndInteraction(M, V, "antagonism")
      M = end[[1]]
      V = end[[2]]     
    }
    
    # if there are no zero lines, stop
    else{
      break
    }
  }
  
  # return the current M and V matrices
  mats = list(M, V)
  return(mats)
}

File: Python/ZeroLines.py
def ZeroLines(M, V, n_sp, antprob):
    '''Function to check for zero lines in the adjacency matrices.
    If there are some zeroed lines, the Coevolution simulation will stop
    Args:
        M: Adjacency matrix of interactions with 0's
        V: Adjacency matrix of antagonistic interactions with 1's
        n_sp: number of species in the matrix
    antprob: Probabilibty of a link become antagonist
    Return
        new M and V matrices with non-zero lines'''

    import numpy as np
    from Antagonize import Antagonize
    from EndInteraction import EndInteraction

    for z in range(0, 1000000):

        m = M.sum(axis = 1)

        if 0 in m:
            # create M and V matrices
            M = np.ones((n_sp, n_sp))
            np.fill_diagonal(M, 0)
            # antagonize M and generate V matrix
            antagonize = Antagonize(M, antprob)
            M = antagonize[0]
            V = antagonize[1]

            # end the pure antagonism
            end = EndInteraction(M, V, "antagonism")
            M = end[0]
            V = end[1]

        else:
            break

    return M, V